src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.pi.lex
+import data.finsupp.order
+import data.finsupp.ne_locus
+
+/-!
+# Lexicographic order on finitely supported functions
+
+This file defines the lexicographic order on `finsupp`.
+-/
+
+variables {α N : Type*}
+
+namespace finsupp
+
+section N_has_zero
+variables [has_zero N]
+
+/-- The lexicographic relation on `α →₀ N`, where `α` is ordered by `r`,
+  and `N` is ordered by `s`. -/
+protected def lex (r : α → α → Prop) (s : N → N → Prop) (x y : α →₀ N) : Prop :=
+pi.lex r (λ _, s) x y
+
+instance [has_lt α] [has_lt N] : has_lt (lex (α →₀ N)) :=
+⟨λ f g, finsupp.lex (<) (<) (of_lex f) (of_lex g)⟩
+
+instance lex.is_strict_order [linear_order α] [partial_order N] :
+  is_strict_order (lex (α →₀ N)) (<) :=
+let i : is_strict_order (lex (α → N)) (<) := pi.lex.is_strict_order in
+{ irrefl := to_lex.surjective.forall.2 $ λ a, @irrefl _ _ i.to_is_irrefl a,
+  trans := to_lex.surjective.forall₃.2 $ λ a b c, @trans _ _ i.to_is_trans a b c }
+
+variables [linear_order α]
+
+/--  The partial order on `finsupp`s obtained by the lexicographic ordering.
+See `finsupp.lex.linear_order` for a proof that this partial order is in fact linear. -/
+instance lex.partial_order [partial_order N] : partial_order (lex (α →₀ N)) :=
+partial_order.lift (λ x, to_lex ⇑(of_lex x)) finsupp.coe_fn_injective--fun_like.coe_injective
+
+variable [linear_order N]
+
+/-- Auxiliary helper to case split computably. There is no need for this to be public, as it
+can be written with `or.by_cases` on `lt_trichotomy` once the instances below are constructed. -/
+private def lt_trichotomy_rec {P : lex (α →₀ N) → lex (α →₀ N) → Sort*}
```

Comment:
I'll wait if @eric-wieser has an opinion on the matter.  I do not know what happens with `private` definitions, other than that I prefer to avoid them.

If Eric does not have a preference, then I will mark the remaining definitions as `@[irreducible]`.

Reply:
I've marked `@[irreducible]` the two decidable instances.  I did not do it for the linear_order.  Is this what you had in mind?  Should I mark also the linear order as irreducible, but after I proved the lemma with its definition?

I am not sure about what the long-term implications of this are, so I am just following my nose and seeing what happens...

Reply:
This way looks good, the remaining reducible definitions living in `Prop` don't really need to unfold usefully.